x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -107,11 +113,12 @@ public Analyzer(AnalyzerContext context, Verifier verifier) {
     }
 
     public LogicalPlan analyze(LogicalPlan plan) {
-        return verify(execute(plan));
+        BitSet partialMetrics = new BitSet(FeatureMetric.values().length);
+        return verify(execute(plan), gatherPartialMetrics(plan, partialMetrics));
     }
 
-    public LogicalPlan verify(LogicalPlan plan) {
-        Collection<Failure> failures = verifier.verify(plan);
+    public LogicalPlan verify(LogicalPlan plan, BitSet partialMetrics) {
+        Collection<Failure> failures = verifier.verify(plan, partialMetrics);
```

Comment:
I'd only have one note, similar to Costin's: couldn't we gather all the used "features"/bitset here in the Analyzer and increase the metrics into an `else` branch here below (i.e. if there are no failures, which I guess is the reason we do that in the Verifier)?

Reply:
It does make sense, and it adds all the logic in one place, but I found the approach a bit forced. The Analyzer gets a Verifier (which is "singleton"), which in turn has the Metrics "singleton" instance in it. Switching to having all the metrics code in the Analyzer means taking out the Metrics from Verifier and putting it in the Analyzer. It would work, but it didn't feel quite right to create a new Analyzer per EsqlSession and pass the Metrics to the session (in addition to the Verifier). Happy to make it like this, if anyone is in favor.